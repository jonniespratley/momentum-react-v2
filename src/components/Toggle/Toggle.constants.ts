const CLASS_PREFIX = 'md-toggle';

const DEFAULTS = {
  IS_DISABLED: false,
  IS_SELECTED: false,
  LABEL: undefined,
};

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
  toggle: `${CLASS_PREFIX}-switch`,
  label: `${CLASS_PREFIX}-label`,
  disabled: `${CLASS_PREFIX}--disabled`,
  focused: `${CLASS_PREFIX}--focused`,
  on: `${CLASS_PREFIX}--on`,
  off: `${CLASS_PREFIX}--off`,
};

export { CLASS_PREFIX, DEFAULTS, STYLE };
