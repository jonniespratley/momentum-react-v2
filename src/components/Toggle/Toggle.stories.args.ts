import { commonStyles } from '../../storybook/helper.stories.argtypes';

import { TOGGLE_CONSTANTS as CONSTANTS } from '.';

const toggleArgTypes = {
  isSelected: {
    description: 'Whether the toggle is on or off',
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: CONSTANTS.DEFAULTS.IS_SELECTED,
    },
  },
  isDisabled: {
    description: 'Whether the toggle is disabled or not',
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: CONSTANTS.DEFAULTS.IS_DISABLED,
    },
  },
  label: {
    description: 'The label that the toggle should have.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: CONSTANTS.DEFAULTS.LABEL,
    },
  },
};

export { toggleArgTypes };

export default {
  ...commonStyles,
  ...toggleArgTypes,
  onChange: { action: 'selected' },
};
