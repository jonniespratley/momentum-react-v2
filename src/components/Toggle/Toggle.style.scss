@use "sass:math";

.md-toggle-wrapper {
  $switch-width: 2.75rem;
  $switch-height: math.div($switch-width, 2);
  $thumb-size: 0.635rem;
  $thumb-margin: math.div($switch-height - $thumb-size, 2);
  $transition-time: 350ms;

  display: flex;
  align-items: center;
  font-family: CiscoSansTT;
  font-family: var(--md-globals-font-default);
  width: fit-content;

  .md-toggle-switch {
    width: $switch-width;
    height: $switch-height;
    border-radius: math.div($switch-height, 2);
    transition: background $transition-time;
    margin-right: 0.5rem;

    &:after {
      content: '';
      width: $thumb-size;
      height: $thumb-size;
      border-radius: math.div($thumb-size, 2);
      margin: $thumb-margin;
      display: block;
      transition: transform $transition-time;
    }

    &.md-toggle--on:after {
      transform: translateX($switch-width - $switch-height);
    }

    &.md-toggle--focused {
      box-shadow: var(--md-globals-focus-ring-box-shadow);
    }

    // Base
    &.md-toggle--off {
      background-color: var(--togglemomentumstyle-untoggled-background);

      &:after {
        background-color: var(--togglemomentumstyle-untoggled-thumb-tint);
      }
    }

    &.md-toggle--on {
      background-color: var(--togglemomentumstyle-toggled-background);

      &:after {
        background-color: var(--togglemomentumstyle-toggled-thumb-tint);
        transform: translateX($switch-width - $switch-height);
      }
    }

    // Hovered
    &:hover {
      &.md-toggle--off {
        background-color: var(--togglemomentumstyle-untoggled-background-hovered);

        &:after {
          background-color: var(--togglemomentumstyle-untoggled-thumb-tint-hovered);
        }
      }

      &.md-toggle--on {
        background-color: var(--togglemomentumstyle-toggled-background-hovered);

        &:after {
          background-color: var(--togglemomentumstyle-toggled-thumb-tint-hovered);
          transform: translateX($switch-width - $switch-height);
        }
      }
    }

    // Pressed
    &:active {
      &.md-toggle--off {
        background-color: var(--togglemomentumstyle-untoggled-background-pressed);

        &:after {
          background-color: var(--togglemomentumstyle-untoggled-thumb-tint-pressed);
        }
      }

      &.md-toggle--on {
        background-color: var(--togglemomentumstyle-toggled-background-pressed);

        &:after {
          background-color: var(--togglemomentumstyle-toggled-thumb-tint-pressed);
          transform: translateX($switch-width - $switch-height);
        }
      }
    }

    // Disabled
    &.md-toggle--disabled {
      &.md-toggle--off {
        background-color: var(--togglemomentumstyle-untoggled-background-disabled);

        &:after {
          background-color: var(--togglemomentumstyle-untoggled-thumb-tint-disabled);
        }
      }

      &.md-toggle--on {
        background-color: var(--togglemomentumstyle-toggled-background-disabled);

        &:after {
          background-color: var(--togglemomentumstyle-toggled-thumb-tint-disabled);
          transform: translateX($switch-width - $switch-height);
        }
      }
    }
  }
}
