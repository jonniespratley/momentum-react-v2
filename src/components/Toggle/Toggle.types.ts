import { CSSProperties } from 'react';
import { AriaSwitchBase } from '@react-types/switch';

export interface Props extends Omit<AriaSwitchBase, 'children'> {
  /**
   * Custom class for overriding this component's CSS.
   */
  className?: string;

  /**
   * Custom id for overriding this component's CSS.
   */
  id?: string;

  /**
   * Custom style for overriding this component's CSS.
   */
  style?: CSSProperties;

  /**
   * Boolean to describe when toggle is on/off
   */
  isSelected?: boolean;

  /**
   * String that displays the label of this toggle.
   */
  label?: string;
}
