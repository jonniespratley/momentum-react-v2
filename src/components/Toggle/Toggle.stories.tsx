import { MultiTemplate, Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';

import Toggle, { ToggleProps, TOGGLE_CONSTANTS } from './';
import Documentation from './Toggle.stories.docs.mdx';
import argTypes from './Toggle.stories.args';

export default {
  title: 'Momentum UI/Toggle',
  component: Toggle,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
};

// NOTE: Primary story. This renders a single component with all external props.
const Example = Template<ToggleProps>(Toggle).bind({});

Example.argTypes = { ...argTypes };

Example.args = {
  defaultSelected: TOGGLE_CONSTANTS.DEFAULTS.IS_SELECTED,
  isDisabled: TOGGLE_CONSTANTS.DEFAULTS.IS_DISABLED,
  label: 'Example text',
};

// NOTE: Common variants story. This renders multiple variants of a single component.
const Common = MultiTemplate<ToggleProps>(Toggle).bind({});

Common.argTypes = { ...argTypes };
delete Common.argTypes.defaultSelected;
delete Common.argTypes.isDisabled;
delete Common.argTypes.label;
delete Common.argTypes.onChange;

Common.args = {};
Common.parameters = {
  variants: [
    { defaultSelected: true, isDisabled: false, label: 'Selected' },
    { defaultSelected: true, isDisabled: true, label: 'Selected + Disabled' },
    { defaultSelected: false, isDisabled: false, label: 'Not selected' },
    { defaultSelected: false, isDisabled: true, label: 'Not selected + Disabled' },
  ],
};

// NOTE: Export stories here. The first export should be `Example`, and the last export should be `Common`.
export { Example, Common };
