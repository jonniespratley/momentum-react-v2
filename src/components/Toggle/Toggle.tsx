import React, { FC } from 'react';
import classnames from 'classnames';
import { VisuallyHidden } from '@react-aria/visually-hidden';
import { useSwitch } from '@react-aria/switch';
import { useToggleState } from '@react-stately/toggle';
import { useFocusRing } from '@react-aria/focus';

import { STYLE } from './Toggle.constants';
import { Props } from './Toggle.types';
import './Toggle.style.scss';

/**
 * The Toggle component. Also known as Switch.
 */
const Toggle: FC<Props> = (props: Props) => {
  const { id, className, style, isDisabled, label } = props;

  const ref = React.useRef();

  const children = label && <span className={STYLE.label}>{label}</span>;
  const ariaProps = { ...props, children };

  const state = useToggleState(ariaProps);
  const { inputProps } = useSwitch(ariaProps, state, ref);
  const { isFocusVisible, focusProps } = useFocusRing();

  return (
    <label
      className={classnames(className, STYLE.wrapper)}
      id={id}
      style={style}
      data-disabled={!!isDisabled}
      data-selected={state.isSelected}
    >
      <VisuallyHidden>
        <input {...inputProps} {...focusProps} ref={ref} />
      </VisuallyHidden>
      <div
        className={classnames(STYLE.toggle, {
          [STYLE.on]: state.isSelected,
          [STYLE.off]: !state.isSelected,
          [STYLE.disabled]: isDisabled,
          [STYLE.focused]: isFocusVisible,
        })}
      />
      {children}
    </label>
  );
};

export default Toggle;
