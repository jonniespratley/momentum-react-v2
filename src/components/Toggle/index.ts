import { default as Toggle } from './Toggle';
import * as CONSTANTS from './Toggle.constants';
import { Props } from './Toggle.types';

export { CONSTANTS as TOGGLE_CONSTANTS };

export type ToggleProps = Props;

export default Toggle;
