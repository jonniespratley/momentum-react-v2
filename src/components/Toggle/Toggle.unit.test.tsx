import React from 'react';
import { mount } from 'enzyme';

import Toggle, { TOGGLE_CONSTANTS as CONSTANTS } from './';

describe('<Toggle />', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      expect.assertions(1);

      const container = mount(<Toggle aria-label={'toggle'} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with label', () => {
      expect.assertions(1);

      const container = mount(<Toggle aria-label={'toggle'} label={'Hello toggle'} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with className', () => {
      expect.assertions(1);

      const className = 'example-class';

      const container = mount(<Toggle className={className} aria-label={'toggle'} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with id', () => {
      expect.assertions(1);

      const id = 'example-id';

      const container = mount(<Toggle id={id} aria-label={'toggle'} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const container = mount(<Toggle style={style} aria-label={'toggle'} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with disabled', () => {
      expect.assertions(1);

      const container = mount(<Toggle isDisabled={true} aria-label={'toggle'} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with selected', () => {
      expect.assertions(1);

      const container = mount(<Toggle isSelected={true} aria-label={'toggle'} />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have its wrapper class adn data atts reflect the default values', () => {
      expect.assertions(3);

      const wrapper = mount(<Toggle aria-label={'toggle'} />)
        .find(Toggle)
        .getDOMNode();

      expect(wrapper.classList.contains(CONSTANTS.STYLE.wrapper)).toBe(true);
      expect(wrapper.getAttribute('data-disabled')).toBe('false');
      expect(wrapper.getAttribute('data-selected')).toBe('false');
    });

    it('should have provided class when className is provided', () => {
      expect.assertions(1);

      const className = 'example-class';

      const element = mount(<Toggle className={className} aria-label={'toggle'} />)
        .find(Toggle)
        .getDOMNode();

      expect(element.classList.contains(className)).toBe(true);
    });

    it('should have provided id when id is provided', () => {
      expect.assertions(1);

      const id = 'example-id';

      const element = mount(<Toggle id={id} aria-label={'toggle'} />)
        .find(Toggle)
        .getDOMNode();

      expect(element.id).toBe(id);
    });

    it('should have provided style when style is provided', () => {
      expect.assertions(1);

      const style = { color: 'pink' };
      const styleString = 'color: pink;';

      const element = mount(<Toggle style={style} aria-label={'toggle'} />)
        .find(Toggle)
        .getDOMNode();

      expect(element.getAttribute('style')).toBe(styleString);
    });

    it('should have provided disabled when disabled is provided', () => {
      expect.assertions(2);

      const component = mount(<Toggle isDisabled={true} aria-label={'toggle'} />);
      const wrapper = component.find(Toggle).getDOMNode();
      const input = component.find('input').getDOMNode();

      expect(wrapper.getAttribute('data-disabled')).toBe('true');
      expect(input.hasAttribute('disabled')).toBe(true);
    });

    it('should have provided selected when selected is provided', () => {
      expect.assertions(2);

      const component = mount(<Toggle isSelected={true} aria-label={'toggle'} />);
      const wrapper = component.find(Toggle).getDOMNode();
      const input = component.find('input').getDOMNode();

      expect(wrapper.getAttribute('data-selected')).toBe('true');
      expect(input.hasAttribute('checked')).toBe(true);
    });
  });

  describe('actions', () => {
    it('should handle mouse press events', () => {
      expect.assertions(4);

      const mockCallback = jest.fn();

      const component = mount(
        <Toggle isSelected={false} onChange={mockCallback} aria-label={'toggle'} />
      ).find('input');

      component.simulate('change', { target: { checked: true } });
      expect(mockCallback).toBeCalledTimes(1);
      expect(mockCallback).toBeCalledWith(true);

      component.simulate('change', { target: { checked: false } });
      expect(mockCallback).toBeCalledTimes(2);
      expect(mockCallback).toBeCalledWith(false);
    });
  });
});
