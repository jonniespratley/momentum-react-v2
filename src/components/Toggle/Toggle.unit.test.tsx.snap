// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> snapshot should match snapshot 1`] = `
<Toggle
  aria-label="toggle"
>
  <label
    className="md-toggle-wrapper"
    data-disabled={false}
    data-selected={false}
  >
    <VisuallyHidden>
      <div
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "clipPath": "inset(50%)",
            "height": 1,
            "margin": "0 -1px -1px 0",
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": 1,
          }
        }
      >
        <input
          aria-checked={false}
          aria-label="toggle"
          checked={false}
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onDragStart={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="switch"
          type="checkbox"
        />
      </div>
    </VisuallyHidden>
    <div
      className="md-toggle-switch md-toggle--off"
    />
  </label>
</Toggle>
`;

exports[`<Toggle /> snapshot should match snapshot with className 1`] = `
<Toggle
  aria-label="toggle"
  className="example-class"
>
  <label
    className="example-class md-toggle-wrapper"
    data-disabled={false}
    data-selected={false}
  >
    <VisuallyHidden>
      <div
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "clipPath": "inset(50%)",
            "height": 1,
            "margin": "0 -1px -1px 0",
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": 1,
          }
        }
      >
        <input
          aria-checked={false}
          aria-label="toggle"
          checked={false}
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onDragStart={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="switch"
          type="checkbox"
        />
      </div>
    </VisuallyHidden>
    <div
      className="md-toggle-switch md-toggle--off"
    />
  </label>
</Toggle>
`;

exports[`<Toggle /> snapshot should match snapshot with disabled 1`] = `
<Toggle
  aria-label="toggle"
  isDisabled={true}
>
  <label
    className="md-toggle-wrapper"
    data-disabled={true}
    data-selected={false}
  >
    <VisuallyHidden>
      <div
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "clipPath": "inset(50%)",
            "height": 1,
            "margin": "0 -1px -1px 0",
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": 1,
          }
        }
      >
        <input
          aria-checked={false}
          aria-label="toggle"
          checked={false}
          disabled={true}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onDragStart={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="switch"
          type="checkbox"
        />
      </div>
    </VisuallyHidden>
    <div
      className="md-toggle-switch md-toggle--off md-toggle--disabled"
    />
  </label>
</Toggle>
`;

exports[`<Toggle /> snapshot should match snapshot with id 1`] = `
<Toggle
  aria-label="toggle"
  id="example-id"
>
  <label
    className="md-toggle-wrapper"
    data-disabled={false}
    data-selected={false}
    id="example-id"
  >
    <VisuallyHidden>
      <div
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "clipPath": "inset(50%)",
            "height": 1,
            "margin": "0 -1px -1px 0",
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": 1,
          }
        }
      >
        <input
          aria-checked={false}
          aria-label="toggle"
          checked={false}
          disabled={false}
          id="example-id"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onDragStart={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="switch"
          type="checkbox"
        />
      </div>
    </VisuallyHidden>
    <div
      className="md-toggle-switch md-toggle--off"
    />
  </label>
</Toggle>
`;

exports[`<Toggle /> snapshot should match snapshot with label 1`] = `
<Toggle
  aria-label="toggle"
  label="Hello toggle"
>
  <label
    className="md-toggle-wrapper"
    data-disabled={false}
    data-selected={false}
  >
    <VisuallyHidden>
      <div
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "clipPath": "inset(50%)",
            "height": 1,
            "margin": "0 -1px -1px 0",
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": 1,
          }
        }
      >
        <input
          aria-checked={false}
          aria-label="toggle"
          checked={false}
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onDragStart={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="switch"
          type="checkbox"
        />
      </div>
    </VisuallyHidden>
    <div
      className="md-toggle-switch md-toggle--off"
    />
    <span
      className="md-toggle-label"
    >
      Hello toggle
    </span>
  </label>
</Toggle>
`;

exports[`<Toggle /> snapshot should match snapshot with selected 1`] = `
<Toggle
  aria-label="toggle"
  defaultSelected={true}
>
  <label
    className="md-toggle-wrapper"
    data-disabled={false}
    data-selected={true}
  >
    <VisuallyHidden>
      <div
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "clipPath": "inset(50%)",
            "height": 1,
            "margin": "0 -1px -1px 0",
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": 1,
          }
        }
      >
        <input
          aria-checked={true}
          aria-label="toggle"
          checked={true}
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onDragStart={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="switch"
          type="checkbox"
        />
      </div>
    </VisuallyHidden>
    <div
      className="md-toggle-switch md-toggle--on"
    />
  </label>
</Toggle>
`;

exports[`<Toggle /> snapshot should match snapshot with style 1`] = `
<Toggle
  aria-label="toggle"
  style={
    Object {
      "color": "pink",
    }
  }
>
  <label
    className="md-toggle-wrapper"
    data-disabled={false}
    data-selected={false}
    style={
      Object {
        "color": "pink",
      }
    }
  >
    <VisuallyHidden>
      <div
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "clipPath": "inset(50%)",
            "height": 1,
            "margin": "0 -1px -1px 0",
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": 1,
          }
        }
      >
        <input
          aria-checked={false}
          aria-label="toggle"
          checked={false}
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onDragStart={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="switch"
          type="checkbox"
        />
      </div>
    </VisuallyHidden>
    <div
      className="md-toggle-switch md-toggle--off"
    />
  </label>
</Toggle>
`;
